drv/lpc55-update-server/src/images.rs

```
@@ -26,79 +45,494 @@ extern "C" {
+    _vector_table_0: [u32; 6],          // 0x08, 0c, 10, 14, 18, 1c
+    nxp_image_length: u32,              // 0x20
+    nxp_image_type: u32,                // 0x24
+    nxp_offset_to_specific_header: u32, // 0x28
+    _vector_table_1: [u32; 2],          // 0x2c, 0x30
+    nxp_image_executation_address: u32, // 0x32
+                                        // Additional trailing vectors are not
+                                        // interesting here.
+                                        // _vector_table_2[u32; 2]
+}
+
+impl ImageVectorsLpc55 {
+    const IMAGE_TYPE_PLAIN_SIGNED_XIP_IMAGE: u32 = 4;
+
+    pub fn is_image_type_signed_xip(&self) -> bool {
+        self.nxp_image_type == Self::IMAGE_TYPE_PLAIN_SIGNED_XIP_IMAGE
+    }
 
     // This part aliases flash in two positions that differ in bit 28. To allow
     // for either position to be used in new images, we clear bit 28 in all of
     // the numbers used for comparison below, by ANDing them with this mask:
-    const ADDRMASK: u32 = !(1 << 28);
+    pub fn normalized_initial_pc(&self) -> u32 {
+        const ADDRMASK: u32 = !(1 << 28);
+        self.initial_pc & ADDRMASK
+    }
+
+    // Length of image from offset zero to end of the signature block (without padding)
+    // a.k.a. ImageVectorsLpc55.nxp_image_length
+    pub fn image_length(&self) -> Option<u32> {
+        if self.is_image_type_signed_xip() {
+            Some(self.nxp_image_length)
+        } else {
+            None
+        }
+    }
+
+    /// Image length padded to nearest page size.
+    pub fn padded_image_len(&self) -> Option<u32> {
+        round_up_to_flash_page(self.image_length()?)
+    }
+
+    /// Determine the bounds of an image assuming the given flash bank
+    /// addresses.
+    pub fn padded_image_range(&self, exec: &Range<u32>) -> Option<Range<u32>> {
+        let image_start = exec.start;
+        let image_end = exec.start.saturating_add(self.padded_image_len()?);
```

Note that this will silently ignore cases where start+len overflows. If saturating arithmetic is correct here, please add a comment explaining why. (I expected `checked_add` to validate the image.)